### DISCUSSION TITLE
Sound Output

### DISCUSSION THREAD

**Discussion by midimusic** [GrandOrgue/grandorgue#1137] [General]
Looking at the GO menu options for sound output GO appears to be using both PortAudio AND RtAudio. By default all "Sound port" options are ticked even for options the system does not have installed. This cannot be right.

 Each of these two systems provides a comprehensive choice of outputs (depending on what is installed on the system). Thus I do not understand why both are needed.

The "Mapping output" menu  appears to be making an arbitrary selection of one sound output system and has no option to select the other. In the system I am looking at now it selects PortAudio with no possibility of selecting RtAudio even if "Pa" is unticked. Unticking Pa simply gives an error saying it is not available even though Rt Audio is ticked and available.

This situation is illogical and I find it a bit weird - Any comments?

csw900

**Comment by oleg68** [comment]
> Looking at the GO menu options for sound output GO appears to be using both PortAudio AND RtAudio. 

GO also supports Jack sound port.

> By default all "Sound port" options are ticked even for options the system does not have installed. This cannot be right.

Both PortAudio and RtAudio are built-in to GrandOrgue executables so they aren't needed to be installed.

> Each of these two systems provides a comprehensive choice of outputs (depending on what is installed on the system). Thus I do not understand why both are needed.

They support different audio drivers. You may use Rt or PA depending on what audio drivers you have installed.

>  it selects PortAudio with no possibility of selecting RtAudio even if "Pa" is unticked. Unticking Pa simply gives an error saying it is not available even though Rt Audio is ticked and available.

Click to Change for switching between different audio devices

**Comment by midimusic** [comment]
Thanks Oleg for the clues.

I would be interested to know EXACTLY what USEFUL drivers you think RtAudio has that PortAudio does not. Both of them work equally well for me on both Windows and Linux. PortAudio is the default selection on both systems. So I think it would simplify things a lot if RtAudio was removed from GO. As far as I can tell it adds nothing apart from making audio setup more difficult.

Also things would be greatly simplified if the audio setup polled the system to determine exactly what audio drivers were AVAILABLE and list ONLY those so the users could select the one they wanted to use. This would be much better than making the user choose from a whole load of drivers, most of which are probably not available or have no beneficial value.

Similarly it would be useful if ONLY the available Sound Devices were listed for the user's choice.

Have you considered using libsoundio?
![portaudio-external-architecture-diagram](https://user-images.githubusercontent.com/6420472/172051466-20a4c271-c079-4c17-b7dc-b8e338a10e74.png)

PS. Having now looked at the documentation, code and api of both systems and run some of their test programs - I am not so keen on portaudio. However I still think that one of them should be removed. I also think that GO's audio setup leaves a lot to be desired and could be significantly improved.

csw900
